amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpSourceStage.scala

```diff
@@ -12,6 +14,21 @@ import com.rabbitmq.client._
 import scala.collection.mutable
 
 final case class IncomingMessage(bytes: ByteString, envelope: Envelope, properties: BasicProperties)
+final case class CommittableIncomingMessage(
+    message: IncomingMessage,
+    callback: AsyncCallback[Unit] = new AsyncCallback[Unit] {
+      override def invoke(t: Unit): Unit = Unit
+    }
+)(implicit channel: Channel) {
+  def ack(multiple: Boolean = false): Unit = {
+    channel.basicAck(message.envelope.getDeliveryTag, multiple)
```

Comment:
This is tricky:

Channel is an internal resource of the stage, so it shouldn't be leaked outside, what if the stage was stopped or failed? Also, I think `basicAck` leads to a network call, is it blocking, if so what thread is this executed on?

On the other hand, if the logic is moved into the stage, then this ack needs to return a `Future[...]` to signal to the stream that the ack was completed or failed before emitting the message downstream.

Reply:
I see. So you are suggesting that I wrapped the Async callback in a plain function and pass that to the Committable message, right?

There is some logic in the Stage so it won't be completed until all the messages have been acked. If something really goes wrong I guess the messages won't be acked. I think it's the same in the akka-streams-kafka, we offer out of the box at-least-once delivery but not exactly-once delivery.

The ack method from rabbitMQ is blocking AFAIK. I tried wrapping it in a Future and using mapAsync but got the issue that I mentioned in a comment above. Maybe you have any suggestion?